pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -236,8 +239,14 @@ func (c *Reconciler) ReconcileKind(ctx context.Context, pr *v1beta1.PipelineRun)
 	if pr.Status.StartTime != nil {
 		// Compute the time since the task started.
 		elapsed := c.Clock.Since(pr.Status.StartTime.Time)
-		// Snooze this resource until the timeout has elapsed.
-		return controller.NewRequeueAfter(pr.PipelineTimeout(ctx) - elapsed)
+		// Snooze this resource until the appropriate timeout has elapsed.
+		waitTime := pr.PipelineTimeout(ctx) - elapsed
+		if pr.Status.FinallyStartTime == nil && pr.TasksTimeout() != nil {
+			waitTime = pr.TasksTimeout().Duration - elapsed
+		} else if pr.FinallyTimeout() != nil {
```

Comment:
this logic might not be quite right if the time until a Pipeline timeout is less than the time until a finally timeout. Ex: Pipeline timeout is 10 minutes, finally timeout is 9 minutes, and finally tasks started 5 minutes into execution.

Reply:
Good catch - it should always be the smallest. Fixing.